File: sast-python2/case/completeness/single_app_tracing/function_call/override/polymorphism_override_002_F.py
# -*- coding: utf-8 -*-
# evaluation information start
# real case = false
# evaluation item = 完整度->单应用跟踪完整度->函数和方法调用->方法重写
# scene introduction = 多态
# level = 2
# bind_url = completeness/single_app_tracing/function_call/override/polymorphism_override_002_F
# evaluation information end
import os


def polymorphism_override_002_F(taint_src):

    class Base(object):
     def call(self):
        raise NotImplementedError(u"Subclasses must implement this method")


    class Sub1(Base):
     def call(self):
        return taint_src


    class Sub2(Base):
     def call(self):
        return u"_"

    sub = Sub2()
    taint_sink(sub.call())


def taint_sink(o):
    os.system(o)


if __name__ == u'__main__':
    taint_src = u"taint_src_value"
    polymorphism_override_002_F(taint_src)

File: sast-go/cases/completeness/single_app_tracing/function_call/override/polymorphism_override_002_F/polymorphism_override_002_F.go
// evaluation information start
// real case = false
// evaluation item = 完整度->单应用跟踪完整度->函数和方法调用->方法重写
// scene introduction = 多态
// level = 2
// bind_url = completeness/single_app_tracing/function_call/override/polymorphism_override_002_F/polymorphism_override_002_F
// evaluation information end

package polymorphism_override_002_F

func polymorphism_override_002_F(__taint_src interface{}) {
	var sub Base
	sub = Sub2{} // 将Sub1实例化
	__taint_sink(sub.call(__taint_src))
}

type Base interface {
	call(src interface{}) interface{}
}

type Sub1 struct{}

func (s Sub1) call(src interface{}) interface{} {
	return src
}

type Sub2 struct{}

func (s Sub2) call(src interface{}) interface{} {
	return "_"
}

func __taint_sink(o interface{}) {
}
